### PR TITLE
Fix ClassCastException caused by 2.x's filter in some condition

### PR DESCRIPTION
**Pull request by wegod**
## What is the purpose of the change

When upgrading from dubbo 2.x to 3.x, we have many projects and need to call provider service using 2.x filter and return RpcResult in some case. 
So it's better to compatible of RpcResult when 3.x consumer calling 2.x provider

## Brief changelog
Fix ClassCastException caused by 2.x's filter in some condition and compatible of RpcResult in dubbo version 2.x

## Verifying this change

I test on my local env

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GitHub issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Each commit in the pull request should have a meaningful subject line and body.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Check if is necessary to patch to Dubbo 3 if you are work on Dubbo 2.7
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Add some description to [dubbo-website](https://github.com/apache/dubbo-website) project if you are requesting to add a feature.
- [ ] GitHub Actions works fine on your own branch.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcResult.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    @Override
+    public void setObjectAttachment(String key, Object value) {
+
+    }
+
+    public String toString() {
+        return "RpcResult [result=" + this.result + ", exception=" + this.exception + "]";
+    }
+}
+
```
Comment (AlbumenJ) [inline dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcResult.java:115]:
```suggestion
public class RpcResult extends AppResponse implements com.alibaba.dubbo.rpc.Result {
    public RpcResult() {
    }

    public RpcResult(Object result) {
        super(result);
    }

    public RpcResult(Throwable exception) {
        super(exception);
    }
}

```

**Review by AlbumenJ — CHANGES_REQUESTED** [review]

File: dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcResult.java (hunk 1 above)
Comment (wegod) [reply dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcResult.java]: Thanks for your code review.
I tested on my local env, and add whenCompleteWithContext's empty implement which is called in FilterChainBuilder.

File: dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcResult.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public RpcResult(Throwable exception) {
+        super(exception);
+    }
+    
+    @Override
+    public org.apache.dubbo.rpc.Result whenCompleteWithContext(BiConsumer<org.apache.dubbo.rpc.Result, Throwable> fn) {
+        return null;
+    }
```
Comment (AlbumenJ) [inline dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcResult.java:38]: What is the purpose of adding this?

**Comment by AlbumenJ** [comment]
@wegod I have try to help you fix the test cases :)

File: dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcResult.java (hunk 2 above)
Comment (wegod) [reply dubbo-compatible/src/main/java/com/alibaba/dubbo/rpc/RpcResult.java]: Wait a moment, I found when dubbo3 using dubbo2's Filter, RpcResult will be handle as AsyncRpcResult. 
This function whenCompleteWithContext will be changed later, more code need to fix.

**Comment by wegod** [comment]
> @wegod I have try to help you fix the test cases :)

Thanks very much. 
Right now I'm using 3.1.10.  Will this branch merge into 3.1.11 or 12?

**Comment by wegod** [comment]
@AlbumenJ My test is done, I found new change in Filter.java and CompatibleResult.java that's better than my first code change. 
All case works in both provider side and consumer side, no matter dubbo2.x or 3.x is provider or consumer.

**Comment by AlbumenJ** [comment]
> > @wegod I have try to help you fix the test cases :)
> 
> Thanks very much. Right now I'm using 3.1.10. Will this branch merge into 3.1.11 or 12?

According to the roadmap, 3.1.x is only for security fix now. We can only maintian one branch due to limited working time (:

**Review by chickenlj — APPROVED** [review]
LGTM.

**Pull request merged**